README
perlnow.el is a collection of utility commands for perl programmers
who use emacs. These commands are designed to help automate some
routine tasks in (1) the initial creation of perl scripts and
modules (2) running, testing and debugging inside of emacs.

DOCUMENTATION

Detailed documentation has been included inside the code itself,
and can be accessed using the emacs help system:

   ESC x help v perlnow-documentation-tutorial

Think of it as the poor man's pod (or perhaps the lazy man's info).

INSTALLATION

Put the perlnow.el file somewhere that's included in your load-path.

Also install Christoph Wedler's template.el because many features
of this package depend on it.  The latest version of template.el
can be found at:

   http://sourceforge.net/project/showfiles.php?group_id=47369

You'll fine a set of template files ("*.tpl") you can use here:

  http://github.com/doomvox/perlnow/tree/master/templates/standard/
        
To start with, copy those templates to your templates directory, 
which by default is ~/.templates. 

In your ~/.emacs file, add something like the following:

   (require 'template)
   (template-initialize)
   (require 'perlnow)

   (perlnow-define-standard-keymappings)

Alternately, if you'd like a different prefix than the
default "C-c\", you can supply it as an argument:

   (perlnow-define-standard-keymappings "C-c'")

You might also like to set some customization variables 
like so:

  (setq perlnow-script-location
      (substitute-in-file-name "$HOME/bin"))
  (setq perlnow-pm-location
      (substitute-in-file-name "$HOME/lib"))
  (setq perlnow-dev-location
      (substitute-in-file-name "$HOME/dev"))

Additional documentation can be found in the docstrings of 
dummy variables named perlnow-documentation-*, e.g. 

   ESC x help v perlnow-documentation-installation







CHANGES
Version 0.43, Jun 13 2010: 

Bug fixes with script creation from different contexts 
(e.g. man page, woman page, etc).  

Apr 15 2010:

Added a user feature to display the current default @INC in an emacs buffer:

   perlnow-display-inc-array

Examining @INC in a readable format is something I often want, so why not?

Aug 23 2009: 

 o  Added features that use module-starter to create Module::Build
    based projects.

 o  Defined additional template expansions:

    (>>>FULL_DATE<<<)        August 23, 2009
    (>>>AMERICAN_DATE<<<)    8/23/09

 o  Re-wrote perlnow-get-inc-spot-classic: fixes the generation of relative
    paths between a script and a module it uses.

Smaller changes:

 o  perlnow-check now turns on strict as well as warnings during the
    check phase.

 o  Added feature to silently ignore redundant ".pm" extensions fed
    to perlnow-module. Entering "/tmp/dev/Module::Stuff.pm" to
    longer creates "/tmp/dev/Module/Stuff.pm.pm".

 o  Fixed bug with perlnow-h2xs, which would crash when trying to
    find a string in the generated test file which no longer exists:
    now just moves point to bottom of test file.


Sep  6 2009:

The function perlnow-define-standard-keymappings now allows
you to override the "C-c/" prefix with a different choice, for
example:

  (perlnow-define-standard-keymappings "\C-'")

Mar  7 2008:

Modified perlnow-perlversion-old-to-new to pass
through newer style versions such as "5.8.0" unchanged.


Sep  5 2009: 

Removed dependancy of perlnow-edit-test-file on the marker
(>>>9<<<) in the test file template: now it searches for the "use"
(or "use_ok") line for the module, via the new perlnow-jump-to-use
function.



Sep  7 2010: 

Fixed bug in perlnow-get-package-name-from-man that could yield
false positives on scripts, leading to perlnow-script from a
script context having a very bad, misleading runstring refering to
the context script instead of the new one.


Version 0.45, Jan 20 2017:
 
A bugfix release:

 o  Fixed generated runstrings when using "prove".

 o  With Module::Build::Tiny you need not have a MANIFEST
    file, and this confused the old algorithm that identified
    a cpan-style project, and this in turn could break
    finding the "t" directory from a *.pm file.  This is fixed.

 o  Changed test policy defaults:

    perlnow-test-path                   was: "." "../t" "./t",  now: "../t" "./t"
    perlnow-test-policy-test-location   was: "./t",             now: "../t"
    perlnow-test-policy-dot-definition  was: "fileloc",         now: "incspot"

    (My ideas about good locations to place test files-- and my
    experience about where they're likely to be found-- has
    changed over the years.)
          
 o  Fixed bug where perlnow-run could get confused by a blank
    setting of existing-run-string

 o  New customization variable: perlnow-force.  When this is
    non-nil, perlnow will avoid asking questions about issues
    like directory creation and so on.  (This is useful for
    writing tests, at the very least.)

 o  When multiple applicable test files are available, the guess
    that perlnow makes as to which one you're most likely to
    be interested in now favors a recent manual selection
    (experimental).

 o  Created a new command, perlnow-run-perltidy, which works on
    the entire buffer by default (operating on the region alone
    can confuse perltidy, because it can't know anything about
    surrounding context).   The perltidy commands have some
    added safety features, checking to make sure the perltidy
    script can be found, and avoiding use on non-perl buffers.

  o Removed all code for deprecated "alt" run strings, and
    moved to a new package perlnow-deprecated.el.

  o plus many other miscellanious small fixes.

(I apologize for the continual minor drift in the UI with
each release.  In the near future I hope to stabilize
things to avoid this going forward.)


Version 0.8, July 31, 2017:

With this release, I'm stripping out the now obsolete
"test-policy" features.

Version 1.0, Oct 10, 2017: 

   Added a suite of tests (written in elisp, using test-simple.el). 

   A major revision in default locations used for file creation.    
                                                                    
   Added Support for milla (a variant of dzil) cpans-style projects.

   Programmatic template lookup that gradually relaxes constraints until an appropriate template is found.

   In addition to 'perlnow-edit-test-file' there's a 'perlnow-test-create' that won't search for existing 
   tests first, and a 'perlnow-test-create-manually' that prompts you to choose a name for the new test.
   The standard keymap has these bound to "a" and "A" respectively.                                     

   New navigation features to skip to next/prev sub ("n" and "p"),
   or the next/prev TODO comment ("N" and "P").

   Added variants of 'narrow-to-defun': 'perlnow-narrow-to-defun',
   'perlnow-narrow-to-defun-other-buffer'.  These include a block 
   of pod before a sub as part of the sub.

   Changed default naming conventions for *.t files, e.g. 
   *.t for scripts have a "-script" suffix, *.t files for 
   modules include the current subname.

Version 1.01, November 27, 2018:
       
   Test creation from a module now regards the top of the file as
   a zone without an associated sub (previously it used the first
   sub)-- this facilitates creating general *.t files for an 
   entire module.



t/test-init.el
;;; test-init.el ---

;; Copyright 2017 Joseph Brenner
;;
;; Author: doom@kzsu.stanford.edu
;; Version: $Id: TEMPLATE.el.tpl,v 1.3 2016/10/21 17:38:51 doom Exp $
;; Keywords:
;; X-URL: not distributed yet
;; License: the same as your GNU emacs (see below)

;;; Commentary:

;; This is essentially an include file, to be called from an
;; elisp *.t file like so:

;;   (if (file-exists-p "test-init.el")
;;      (load-file "test-init.el"))

;; This file is a place to put shared set-up customizations for
;; the elisp *.t files in the same directory.

(provide 'test-init)
(require 'cl-lib)

;; general purpose file-system utility
(defun test-init-fixdir (dir &optional root)
  "Fixes the DIR.
Conditions directory paths for portability and robustness.
Some examples:
 '~/tmp'             => '/home/doom/tmp/'
 '~/tmp/../bin/test' => '/home/doom/bin/test/'
Note: converts relative paths to absolute, using the current
default-directory setting, unless specified otherwise with the
ROOT option.  Note side-effect: converts the empty string into
the default-directory or the ROOT setting."
  (let ((return
         (substitute-in-file-name
          (convert-standard-filename
           (file-name-as-directory
            (expand-file-name dir root))))))
    return))

;; load-path additions
(setq load-path (cons
                 (test-init-fixdir "$HOME/lib/emacs/local")
                 load-path))
(setq load-path (cons
                 (test-init-fixdir "$HOME/End/Cave/Perlnow/lib/perlnow/elisp/")
                 load-path))
(setq load-path (cons
                 (test-init-fixdir "$HOME/End/Sys/Emacs/emacs-test-simple/")
                 load-path))

(defconst test-init-slash (convert-standard-filename "/")
  "A more portable form of the file system name separator.")

;; TODO why not the full test-init-* prefix on the following vars?
(defvar test-bin
  (file-name-directory
   (cond (load-file-name load-file-name)
         (t
          (concat test-init-slash "tmp")
          )))
  "Location of this test script.
Falls back to unix /tmp if not run as script.")

(defvar test-data
  (file-name-as-directory
   (concat test-bin "dat")))

(defvar test-loc
  (file-name-as-directory
   (substitute-in-file-name
    (concat "$HOME" test-init-slash
            "tmp"   test-init-slash
            "perlnow_test")))
  "Root location for perlnow tests: a tree of scratch directories.")

(defvar test-init-loc-isolated ""
  "Full path to a specific scratch directory for the current tests.")

(require 'test-simple)
(require 'template)
(template-initialize)
(require 'perlnow)

(setenv "USE_TAP" "t")

;;========
;; perlnow-specific functions

(defun test-init ()
  "Initialize a perlnow test. Wrapper around test-init-standard."
  (setq perlnow-quiet t) ;; ask me no questions
  ;; (perlnow-tron)
  (let ((loc (test-init-standard)) )
    loc))

(defun test-init-setup-perlnow-locations ( sub-directory )
  "Sets up standard perlnow locations using the given SUB-DIRECTORY.
Creates the SUB-DIRECTORY in the `test-loc' location, and adds
the tree of standard perlnow locations (bin, lib, dev, etc), setting
the perlnow customization variables to those locations.
NOTE: if SUB-DIRECTORY already exists it is deleted first
with `test-init-safe-recursive-delete'.
The goal here is isolated tests without side-effects on each
other.
Returns the full-path to the new sub-directory."
  (let* ((deep-test-loc
          (test-init-fixdir (concat test-loc sub-directory))))
    (test-init-safe-recursive-delete deep-test-loc)
    (test-init-mkpath deep-test-loc)
    (setq perlnow-script-location
          (file-name-as-directory (concat deep-test-loc "bin")))
    (setq perlnow-pm-location
          (file-name-as-directory (concat deep-test-loc "lib")))
    (setq perlnow-dev-location
          (file-name-as-directory (concat deep-test-loc "dev")))
    (setq perlnow-etc-location
          (file-name-as-directory (concat deep-test-loc "etc")))
    (setq perlnow-incspot-from-t-stash-file
          (concat perlnow-etc-location "incspot_from_t.json"))
    (test-init-mkpath perlnow-script-location)
    (test-init-mkpath perlnow-pm-location)
    (test-init-mkpath perlnow-dev-location)
    (test-init-mkpath perlnow-etc-location)
    (setq test-init-loc-isolated deep-test-loc)
    deep-test-loc))

(defun test-init-standard ()
  "Generates a test tree in a sub-directory named with the script's file-name prefix.
E.g. for 02-check_it.t, creates a \"t02\" in `test-loc' by running
\\[test-init-setup-perlnow-locations].  Returns the full path to the new location."
  (let (test-loc-subdir)
    (setq test-loc-subdir (test-init-standard-core))
    (test-simple-start) ;; Zero counters and start the stop watch.
    test-loc-subdir))

(defun test-init-standard-core ()
  "Implements main work done by \\[test-init-standard].
You might call this directly if you want the test location
without running \\[test-simple-start], but also see `test-init-loc-isolated'."
  (let* ( count-args
          script-file-name
          file-prefix
          sub-directory
          test-loc-subdir )
    (setq count-args (length command-line-args))
    (setq sub-directory
          (cond ((>= count-args 2)
                 (setq script-file-name (nth 2 command-line-args))
                 (setq file-prefix
                     (cond ( (string-match "\\.t$" script-file-name)
                             (car (split-string (file-name-nondirectory script-file-name) "-"))
                            )
                           (t ;; if no numeric prefix, use this alternate string
                            "XX"
                           )))
                 ;; (setq script-loc (file-name-directory script-file-name))
                 (concat "t" file-prefix)
                )
                (t
                 "tXXXX"
                 )))
    (setq test-loc-subdir (test-init-setup-perlnow-locations sub-directory))
    ;; (test-simple-start) ;; Zero counters and start the stop watch.
    test-loc-subdir))


;;========
;; General purpose file-system utilities
;; (TODO move to another *.el package some day?)
;; Note: some others are defined up top so they can be used in defvar's
;;
(defun test-init-mkpath (dir)
  "Create directory DIR (and intervening levels) if it doesn't exist."
  (unless (file-directory-p dir)
     (make-directory dir t)))

(defun test-init-move-file-out-of-way (filename &optional extension)
  "Move FILENAME out of the way, by renaming it with appended EXTENSION.
Default EXTENSION is \".OLD\""
  (let* ((extension (cond (extension extension)
                          (t ".OLD")))
         (backup-name (concat filename extension))
         )
   (cond ((file-exists-p filename)
          (message "renaming existing file: %s as %s" filename backup-name)
          (rename-file filename backup-name t) ;; with t, overwrites
          ))
   ))

;; TODO could you replace this with:
;;   (setq trash-directory "Old")
;;   (move-file-to-trash dirname)
(defun test-init-safe-recursive-delete (dirname &optional backup-location safety-override)
  "Given a DIRNAME including a full-path, move it to the BACKUP-LOCATION.
BACKUP-LOCATION defaults to a sub-directory named \"Old\".
If a directory of this name already exists in the backup-location,
this will delete it first: we preserve only the last version.
As a safety feature, this first checks to make sure that the DIRNAME
contains a word such as 'tmp', 'temp' or 'test', indicating that
it's intended to be ephemeral.  If SAFETY-OVERRIDE is non-nil, this check is skipped"
  (setq dirname (test-init-fixdir dirname))
  (cond ((file-exists-p dirname)
         (let* (
                ;; remove trailing slash so file-name-directory & nondirectory can work
                (last-slash-pat (concat test-init-slash "$"))
                (dirname-trimmed
                 (replace-regexp-in-string last-slash-pat "" dirname))
                (dirname-path      (file-name-directory    dirname-trimmed))
                (dirname-sans-path (file-name-nondirectory dirname-trimmed))

                dirname-path-fixed default-backup-location
                new-backup new-backup-temp )

           ;; bring back trailing slash
           (setq dirname-path-fixed (test-init-fixdir dirname-path))
           (setq default-backup-location  (test-init-fixdir (concat dirname-path-fixed "Old")))

           (unless backup-location
             (setq backup-location default-backup-location))
           (test-init-mkpath backup-location)

           (setq new-backup
                 (test-init-fixdir
                  (concat backup-location dirname-sans-path)))

           ;; Get a uniq directory name to use temporarily
           (setq new-backup-temp new-backup)
           (let ((suffix "A")
                 (count   0))
             (while (file-exists-p new-backup-temp)
               (setq new-backup-temp
                     (concat backup-location (concat suffix (number-to-string count))))
               (setq count (1+ count)) ))

           (cond ((or
                   safety-override
                   (string-match "\\btest\\b" dirname)
                   (string-match "\\btmp\\b"  dirname)
                   (string-match "\\btemp\\b" dirname)
                   )
                  ;; first copy to a dir with unique name, then
                  ;; delete orignal, and rename the copy using
                  ;; original name
                  (copy-directory dirname new-backup-temp nil t t)
                  (delete-directory dirname t)
                  (if (file-directory-p new-backup)
                      (delete-directory new-backup t))
                  (copy-directory new-backup-temp new-backup nil t t)
                  ))))))

;; TODO
;; Might be able to just use this: dired-copy-file-recursive
;; but it is has no documentation.
(defun test-init-copy-tree (this there)
  "Copies the tree of files THIS over to location THERE.
Shells out to the tar command.  If it it looks like THIS is in
THERE aready, does a \\[test-init-safe-recursive-delete] on it
first."
  (unless (file-accessible-directory-p this)
    "The first argument must be an existing directory.")
  (unless (file-accessible-directory-p there)
    "The second argument must be an existing directory.") ;; TODO Letting me run with non-existant there dir?
  (setq this  (test-init-fixdir this))
  (setq there (test-init-fixdir there))
  (let* (
         ;; remove trailing slash so file-name-directory & nondirectory can work
         (tail-slash-pat (concat perlnow-slash "$"))
         (this-trimmed
              (replace-regexp-in-string tail-slash-pat "" this))
         (orig-loc  (file-name-directory this-trimmed))
         (this-name (file-name-nondirectory this-trimmed))
         (basename  (file-name-sans-extension this-name))
         (new-tree (concat there this-name))
         (tar-file (concat basename ".tgz"))
         (tar-program  "tar")
         (tar-args (list "czf" tar-file "."))
         (display-buffname "*test-init*")
         (display-buffer (get-buffer-create display-buffname))
         (save-loc default-directory)
         )
    (if (file-accessible-directory-p new-tree)
        (test-init-safe-recursive-delete new-tree nil t))
    (cd this)
     ;; tar-cmd-1
     (call-process tar-program nil nil display-buffer
                   "czf" tar-file ".")
    (test-init-mkpath there)
    (copy-file tar-file there t) ;; over-write is okay
    (move-file-to-trash tar-file) ;; EXPERIMENTAL

    ;; tar-cmd-2
    (setq save-loc default-directory)
    (cd there)
    (call-process tar-program nil nil display-buffer "xzf" tar-file)
    (cd save-loc)
    ))

(defun test-init-subdirs (loc)
  "A simple listing of sub-directories in the given directory LOC.
Skips hidden directories (names with leading \".\").
Restricts the listing to accessible directories.
Does not include the path.  Returns a sorted list."
  (dolist (dir (directory-files test-set nil "^[^.]"))
    (if (and
         (file-accessible-directory-p dir)
         (not (string-equal dir "meta")))
        (push dir cases)))
  cases)

;; initialize the 'perlnow_test/tNN' location from 'code/sNN':
;; e.g.
;;   /home/doom/End/Cave/Perlnow/lib/perlnow/t/dat/code/s65/*
;; ==>
;;   /home/doom/tmp/perlnow_test/t65
;;
;; full path to "dat" should be in var: test-data
(defun test-init-load-setup-code (dir-basename &optional deep-test-loc)
  "Copies initial version of directory named DIR-BASENAME to test location.
Moves any existing copy of it out of the way.  The test location
DEEP-TEST-LOC defaults to the return from \\[test-init-standard-core],
passed via `test-init-loc-isolated'."
  (unless deep-test-loc (setq deep-test-loc test-init-loc-isolated))
  ;; (message "test-init-loc-isolated: %s" test-init-loc-isolated) ;; /home/doom/tmp/perlnow_test/t65/
  (let* ((slash test-init-slash)
         (data-loc   (test-init-fixdir (concat default-directory "dat" slash)))
         (source-loc (concat data-loc "code" slash dir-basename slash))
;;         (source-loc default-directory)
         )
  ; (message "source-loc: %s" source-loc) ;; /home/doom/End/Cave/Perlnow/lib/perlnow/t/dat/code/s65/

  (test-init-copy-tree source-loc deep-test-loc)
  ))


(defun test-init-load-data (file)
  "Re-load data structure from elisp FILE generated by a pp dump."
  (let ( raw-string  data )
    (unless (file-exists-p file)
      (message "test-init-load-data: file does not exist: %s" file))
    (find-file file)
    (setq raw-string (buffer-substring-no-properties (point-min) (point-max)))
    (setq data (read raw-string))
    (kill-buffer)
    data))

(defun test-init-preserve-data (data file)
  "Save given elisp DATA to the FILE, using a pp dump."
  (let ( raw-string )
   (find-file file)
    ;; TODO what safety feature can I put on this?
    (delete-region (point-min) (point-max))
    (insert (pp-to-string data))
    (basic-save-buffer)
    ))

;; simpler vesion of perlnow-stash-lookup, but evals the val
(defun test-init-plist-lookup (keystr plist-symbol)
  "Look-up string KEYSTR in the plist indicated by PLIST-SYMBOL.
If KEYSTR is nil, returns nil." ;; TODO not sure it matters if KEYSTR is a string
  (let ((ret
         (cond (keystr
                (let* ((raw (lax-plist-get (symbol-value plist-symbol) keystr))
                       (value (eval raw))
                      )
                  value))
               (t
                nil))))
    ret))






;; LICENSE

;; This program is free software; you can redistribute it and/or modify
;; it under the same terms as the version of GNU Emacs you intend to use it with.

;; At present, GNU Emacs is under the GNU General Public License version 3
;; or (at your option) any later version.  This license is as published by
;; the Free Software Foundation.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;;; elisp-test-init.el ends here


t/dat/code/s01/meta/test_cases.el
("non1"
  '("start-file-list"
    '("bin/glowing.pl" "dev/Radioactive/Words.pm" "dev/Radioactive/Bytes.pm")
   "t-list"
    '("01-Radioactive-Bytes.t" "02-Radioactive-Words.t" "03-glowing-script.t"))
 "git1"
  '("start-file-list"
    '("bin/calc_beauty.pl" "lib/Poetx/Math.pm" "lib/Math/Poetry.pm")
   "t-list"
    '("01-Poetx-Math.t"  "02-Math-Poetry.t"  "03-calc_beauty-script.t"))
 "cpan1"
 '("start-file-list"
  '("Stiletto-Heel/script/mac_the_original.pl" "Stiletto-Heel/lib/Stiletto/Senate.pm" "Stiletto-Heel/lib/Stiletto/Heel.pm")
  "t-list"
  '("01-Stiletto-Heel.t" "02-Stiletto-Senate.t" "03-Stiletto-Heel-mac_the_original-script.t")))


elisp/perlnow-deprecated.el
;;; perlnow-deprecated.el --- Sun  January 01, 2017  13:40

;; Copyright 2017 Joseph Brenner
;;
;; Author: doom@kzsu.stanford.edu
;; Version: $Id: TEMPLATE.el.tpl,v 1.3 2016/10/21 17:38:51 doom Exp $
;; Keywords:
;; X-URL: not distributed yet
;; License: the same as your GNU emacs (see LICENSE below)

;;; Commentary:

;;  Code to support deprecated features, moved here from perlnow.el.
;;

;; Put this file into your load-path and the following into your ~/.emacs:
;;   (require 'perlnow-deprecated)

;;; Code:

(provide 'perlnow-deprecated)
(eval-when-compile
  (require 'cl))


;;;========
;;; The "alt" runstring feature.

;;; Now implementing the "alt-run-string" in addition to
;;; the "run-string": having both allows for
;;; having two separate concurrently defined ways of running the
;;; the perl code in the current buffer.  The heuristics for
;;; guessing what run string to use remain identical.
;;; (( TODO -- maybe they will become identical again someday ))

(defvar perlnow-script-alt-run-string nil
  "The alternative run string for perl scripts, used by \\[perlnow-alt-run].

  Warning: Deprecated feature.

Leave this set to nil unless you want to override the heuristics
used by \\[perlnow-set-alt-run-string] to determine the way to test
the current script.  This is a buffer local variable, i.e. it
may be set differently for different files.")
(put 'perlnow-script-alt-run-string  'risky-local-variable t)
(make-variable-buffer-local 'perlnow-script-alt-run-string)

(defvar perlnow-module-alt-run-string nil
  "The alternative run string for perl modules, used by \\[perlnow-alt-run].

   Warning: Deprecated feature.

Leave this set to nil unless you want to override the heuristics
used by \\[perlnow-set-alt-run-string] to determine the way to test
the current script.  This is a buffer local variable, i.e. it
may be set differently for different files.")
(put 'perlnow-module-alt-run-string  'risky-local-variable t)
(make-variable-buffer-local 'perlnow-module-alt-run-string)

(defvar perlnow-alt-run-string nil
  "Tells \\[perlnow-alt-run] how to run the code in a particular file buffer.
This is a buffer local variable which is set by  \\[perlnow-script-alt-run-string],
and this should not typically be set by the user directly.
See `perlnow-script-alt-run-string' and `perlnow-module-alt-run-string' instead.")
(put 'perlnow-alt-run-string  'risky-local-variable t)
(make-variable-buffer-local 'perlnow-alt-run-string)


;;;==========================================================
;;; Intentionally neglected commands (use "alt" runstrings).
;;; Bumping 'em down here where I can forget about them for now.
;;; Reimplement later.
;;; Maybe: work out a way to extend the main routines
;;; to do the "alt" handling?

(defun perlnow-alt-run (alt-run-string)
  "Run the perl code in this file buffer.

   Warning: Deprecated feature.

This uses an interractively set ALT-RUN-STRING determined
from `perlnow-alt-run-string' which may have been set by using
\\[perlnow-set-alt-run-string].  If `perlnow-alt-run-string' is nil,
\\[perlnow-set-alt-run-string] is called automatically.\n
The alt run string can always be changed later by running
\\[perlnow-set-alt-run-string] manually."
  (interactive
   (let (input)
     (if (eq perlnow-alt-run-string nil)
         (setq input (perlnow-set-alt-run-string))
       (setq input perlnow-alt-run-string))
     (list input)
     ))
  (perlnow-run alt-run-string)) ; Note: uses perlnow-run rather than running compile directly


(defun perlnow-set-alt-run-string ()
  "Prompt the user for a new alternative run string for the current buffer.

  Warning: Deprecated feature.

This sets the global variable `perlnow-alt-run-string' that \\[perlnow-alt-run]
will use to run the code in future in the current buffer.
Frequently, the user will prefer to use \\[perlnow-alt-run] and let it
urun this command indirectly if need be; however using this command
directly is necessary to change the alt-run command string later.  \n
From within a program, it's probably best to set some variables
directly, see `perlnow-script-alt-run-string' and `perlnow-module-alt-run-string'.\n
This function uses \\\[perlnow-module-code-p] to see if the code looks like a
module (i.e. does it have a package line), otherwise it
assumes it's a perl script.  The heuristics for setting a default
\"alt\"-run string are identical to those used for setting the
`perlnow-run-string'."
;;; perlnow-set-alt-run-string is a minor variation of perlnow-set-run-string
  (interactive)
  (cond
   ((perlnow-module-code-p)
    ;; set-up a decent default value
    (unless perlnow-module-alt-run-string
      (progn
        (setq perlnow-module-alt-run-string
              (perlnow-guess-module-run-string))))
    ;; ask user the alternative way to run this module (use as default next time)
    (setq perlnow-module-alt-run-string
          (read-from-minibuffer
           "Set the alternative run string for this module: "
           perlnow-module-alt-run-string))
    ;; tell perlnow-alt-run how to do it
    (setq perlnow-alt-run-string perlnow-module-alt-run-string))
   (t  ;;  assume it's a script since it's not a module.
    ;; set-up intelligent default alt run string
    (unless perlnow-script-alt-run-string
      (progn
        (setq perlnow-script-alt-run-string
              (perlnow-guess-script-run-string))
        ))
    ;; ask user the alternative way to run this script (use as default next time)
    (setq perlnow-script-alt-run-string
          (read-from-minibuffer
           "Set the alternative run string for this script: "
           perlnow-script-alt-run-string))
    ;; tell perlnow-alt-run to do it that way
    (setq perlnow-alt-run-string perlnow-script-alt-run-string))))




;;;==========================================================
;;; cheat commands ("cheat" == automatically fix things so checks pass)
;;;

;; TODO complains if you run this from a non *.t buffer...
;; but if it has a *.t file associated, it should switch to there
;; and do the "revise" magic over there.
;; TODO verify that *compilation* shows results from running
;; the *.t, if not, do a run first.
(defun perlnow-revise-test-plan ()
  "Revise the test plan to match the current count of tests.

   DEPRECATED: just use done_testing()

Presumes you've just run the test, and checks the '*compilation*'
buffer to find out what the actual count of tests are, and
modifies the *.t file's plan to match."
;; Note: presumes that perlnow-run uses standard buffer name *compilation*.
  (interactive)
  (let* ((full-file (buffer-file-name))
         (location (file-name-directory full-file))
         (filename (file-name-nondirectory full-file))
         (extension (perlnow-file-extension filename))
         ;; (original-buffer (current-buffer)) ;; needed?
         (plan-count-pattern
          "Looks like you planned \\([0-9]+\\) test.*?but ran \\([0-9]+\\)")
         (planned-tests)
         (actual-tests)
         (planned-tests-pattern))
    (save-excursion
      ;; verify that we're inside a *.t buffer
      (unless (string= extension "t")
        (error "Can only run perlnow-revise-test-plan on a *.t file."))
      (save-excursion
        (set-buffer "*compilation*")
        (goto-char (point-min))
        (cond ( (re-search-forward plan-count-pattern)
                (setq planned-tests (match-string 1))
                (setq actual-tests  (match-string 2))
                )
              (t
               (message "Is the test plan already up-to-date?")))
        ) ;; return to the *.t buffer
      ;; (set-buffer original-buffer) ;; uncomment, if needed

      ;; Skip to top, search down for 'Test::',
      (goto-char (point-min))
      (cond ( (and
               (re-search-forward "Test::")
               (re-search-forward "plan") )
            ;; No good?  Why?  TODO  "\b" doesn't work either
            ;;  (setq planned-testes-pattern (format "\\b" planned-tests "\\b" ))
              (setq planned-testes-pattern planned-tests)
               (cond ((re-search-forward planned-testes-pattern nil t)
                       (replace-match actual-tests nil t))
                     (t
                      (message "Failed to find expected %s after \"plan\"." planned-tests)
                      )))
            (t
             (message "Test::More plan not found."))
            ))))



;;;=======
;;; general snippets


;; (defvar perlnow-screen-height 60
;; "A hack to deal with the fact that I don't know where the function
;; 'screen-height' that I used to use has gone.")
;; Now using frame-height.






;; LICENSE

;; This program is free software; you can redistribute it and/or modify
;; it under the same terms as the version of GNU Emacs you intend to use it with.

;; At present, GNU Emacs is under the GNU General Public License version 3
;; or (at your option) any later version.  This license is as published by
;; the Free Software Foundation.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;;; perlnow-deprecated.el ends here
